### PR TITLE
Flychecker for CUDA using nvcc

### PR DESCRIPTION
**Pull request by KentJames**
This is a basic flychecker for the NVIDIA cuda language. This should work within the standard cuda-mode defined here: https://www.emacswiki.org/emacs/CudaMode

I have my own custom cuda-mode which is an extension of the one above, which I'll release shortly, but the naming is the same so shouldn't affect flycheck.

nvcc abstracts the underlying system compiler that it uses, so the errors  reported by nvcc and picked up by flycheck should be consistent.

### CODE REVIEW

File: flycheck.el [hunk 1]
```diff
@@ -6972,6 +6973,59 @@ See URL `http://stylelint.io/'."
   :error-parser flycheck-parse-stylelint
   :modes (css-mode))
 
+(flycheck-def-args-var flycheck-cuda-args cuda
+  :package-version '(flycheck . "0.01"))
```
Comment (fmdkdd) [inline flycheck.el]: Package version should be `"32"` here and for other options as well (look at the header of flycheck.el).

File: flycheck.el [hunk 2]
```diff
@@ -173,6 +173,7 @@ attention to case differences."
     coq
     css-csslint
     css-stylelint
+    cuda
```
Comment (fmdkdd) [inline flycheck.el]: The usual naming for checkers is `<language>-<tool-name>`, unless there is only one credible checker in practice.  Is `nvcc` the only CUDA checker out there?

File: flycheck.el [hunk 3]
```diff
@@ -6972,6 +6973,59 @@ See URL `http://stylelint.io/'."
...
+  :package-version '(flycheck . "0.01"))
+
+(flycheck-def-option-var flycheck-cuda-include-path nil cuda
+  "A list of include directories for nvcc."
+  :type '(repeat (directory :tag "Include directory"))
+  :safe #'flycheck-string-list-p
+  :package-version '(flycheck . "0.01"))
+
+(flycheck-define-checker cuda
+  "A CUDA C/C++ syntax checker..
```
Comment (fmdkdd) [inline flycheck.el]: `"... using nvcc."` in place of the `".."`.

File: flycheck.el [hunk 4]
```diff
@@ -6972,6 +6973,59 @@ See URL `http://stylelint.io/'."
...
+
+See URL `https://developer.nvidia.com/cuda-llvm-compiler'."
+  :command ("nvcc"
+            "-c" ;; Compile Only
+            (option "-std=" flycheck-cuda-language-standard concat)
+            (option-list "-include" flycheck-cuda-includes)
+            (option-list "-D" flycheck-cuda-definitions concat)
+            (option-list "-I" flycheck-cuda-include-path)
+            ;; Read from standard input
+            source)
```
Comment (fmdkdd) [inline flycheck.el]: `source` is not standard input: look at the docstring for `flycheck-substitute-argument`.  If you want standard input, you specify that explicitly using `:standard-input t`, and usually pass a special argument to the checker (like `-`).

So either use `source` and remove the comment, or look to see if `nvcc` supports standard input and use that (we prefer standard input if available).

**Comment by KentJames** [comment]
Also, sorry for the delay, work is crucifying now. I'll try and action
this and tidy it up this weekend.

File: flycheck.el (hunk 2 above)
Comment (KentJames) [reply flycheck.el]: Yes nvcc is the only CUDA compiler out there. CUDA is a proprietary language made by nvidia, so there aren't any open source alternatives.

OpenCL is the open source counterpart to CUDA. I might write a checker for this in future when I need one.

File: flycheck.el (hunk 4 above)
Comment (KentJames) [reply flycheck.el]: I've removed the comment and kept it as source for now. 

I'll probably add more bells and whistles to this as I get more comfortable with contributing.

**Comment by KentJames** [comment]
I'm struggling to get this to pass. I'm unsure as to why it's complaining about not-documented options when they are documented in languages.rst. 

It's almost like the checker isn't registered, but it is very explicitly registered. Any ideas?

**Comment by KentJames** [comment]
Okay looks like it's all there now. Tests pass, and travis is happy.

File: flycheck.el [hunk 5]
```diff
@@ -6972,6 +6973,59 @@ See URL `http://stylelint.io/'."
   :error-parser flycheck-parse-stylelint
   :modes (css-mode))
 
+(flycheck-def-option-var flycheck-cuda-language-standard nil cuda-nvcc
+  "Our CUDA Language Standard "
```
Comment (fmdkdd) [inline flycheck.el]: Is there a reason the docstring ends with a space and not with a dot `.`?  If not, please use a dot instead, as the first line of a docstring should be a complete sentence.

The same applies to the `flycheck-cuda-includes` option below.

```suggestion
  "Our CUDA Language Standard."
```

File: flycheck.el [hunk 6]
```diff
@@ -6972,6 +6973,59 @@ See URL `http://stylelint.io/'."
...
+  :type '(choice (const :tag "Default standard" nil)
+                 (string :tag "Language standard"))
+  :safe #'stringp
+  :package-version '(flycheck . "0.32"))
```
Comment (fmdkdd) [inline flycheck.el]: The version number should just be `"32"`, not `"0.32"`.  We have eschewed the dot notation some releases ago.

```suggestion
  :package-version '(flycheck . "32"))
```

File: doc/languages.rst [hunk 7]
```diff
@@ -198,6 +198,35 @@ to view the docstring of the syntax checker.  Likewise, you may use
          The cppcheck suppressions file to use via the
          ``--suppressions-list=`` argument.
 
+.. supported-language:: CUDA C/C++
+   :index_as: CUDA
+
+   .. syntax-checker:: cuda-nvcc
+
+      Checks syntax for CUDA C/C++ using the nvcc compiler bundled in the
```
Comment (fmdkdd) [inline doc/languages.rst]: Please turn `nvcc` into a link to its webpage (the same URL as the one you used for the checker definition).

```suggestion
      Checks syntax for CUDA C/C++ using the `nvcc <https://developer.nvidia.com/cuda-llvm-compiler>` compiler bundled in the
```

File: doc/languages.rst [hunk 8]
```diff
@@ -198,6 +198,35 @@ to view the docstring of the syntax checker.  Likewise, you may use
          The cppcheck suppressions file to use via the
          ``--suppressions-list=`` argument.
 
+.. supported-language:: CUDA C/C++
```
Comment (fmdkdd) [inline doc/languages.rst]: The list of languages is sorted alphabetically, so this whole section should appear after CSS rather.

File: flycheck.el [hunk 9]
```diff
@@ -6972,6 +6973,59 @@ See URL `http://stylelint.io/'."
...
+(make-variable-buffer-local 'flycheck-cuda-language-standard)
+
+(flycheck-def-option-var flycheck-cuda-includes nil cuda-nvcc
+  "Our include directories to pass to nvcc "
```
Comment (fmdkdd) [inline flycheck.el]:
```suggestion
  "Our include directories to pass to nvcc."
```

File: flycheck.el [hunk 10]
```diff
@@ -6972,6 +6973,59 @@ See URL `http://stylelint.io/'."
...
+  :type '(repeat (file :tag "Include file"))
+  :safe #'flycheck-string-list-p
+  :package-version '(flycheck . "0.32"))
```
Comment (fmdkdd) [inline flycheck.el]:
```suggestion
  :package-version '(flycheck . "32"))
```

File: flycheck.el [hunk 11]
```diff
@@ -6972,6 +6973,59 @@ See URL `http://stylelint.io/'."
...
+A list of strings to pass to cuda, a la flycheck-clang"
+  :type '(repeat (string :tag "Definitions"))
+  :safe #'flycheck-string-list-p
+  :package-version '(flycheck . "0.32"))
+
+(flycheck-def-option-var flycheck-cuda-include-path nil cuda-nvcc
+  "A list of include directories for nvcc."
+  :type '(repeat (directory :tag "Include directory"))
+  :safe #'flycheck-string-list-p
+  :package-version '(flycheck . "0.32"))
```
Comment (fmdkdd) [inline flycheck.el]:
```suggestion
  :package-version '(flycheck . "32"))
```

File: flycheck.el [hunk 12]
```diff
@@ -6972,6 +6973,59 @@ See URL `http://stylelint.io/'."
...
+  :type '(repeat (file :tag "Include file"))
+  :safe #'flycheck-string-list-p
+  :package-version '(flycheck . "0.32"))
+
+(flycheck-def-option-var flycheck-cuda-definitions nil cuda-nvcc
+  "Additional preprocessor definitions for nvcc.
+A list of strings to pass to cuda, a la flycheck-clang"
+  :type '(repeat (string :tag "Definitions"))
+  :safe #'flycheck-string-list-p
+  :package-version '(flycheck . "0.32"))
```
Comment (fmdkdd) [inline flycheck.el]:
```suggestion
  :package-version '(flycheck . "32"))
```

**Comment by KentJames** [comment]
I'm a bit mystified why Travis didn't pass that? The lint-python test threw up loads of errors in files I haven't touched. But it passed fine last time... makes no sense.

**Comment by fmdkdd** [comment]
Rebased and merged in #1554.  Thanks @KentJames!

**Pull request closed without merging**